File: Sources/Gantry/Services/CertificatePinStore.swift
import CryptoKit
import Foundation

final class CertificatePinStore: @unchecked Sendable {
    static let shared = CertificatePinStore()

    enum ValidationResult: Equatable {
        case firstUse
        case matched
        case mismatch

        var accepted: Bool { self != .mismatch }
    }

    private let lock = NSLock()
    private let storageKey = "printer-certificate-pins-v1"

    func validate(certificateData: Data, for serial: String) -> ValidationResult {
        let fingerprint = SHA256.hash(data: certificateData).map { String(format: "%02x", $0) }.joined()
        lock.lock()
        defer { lock.unlock() }

        let defaults = BambuDefaults.shared
        var pins = defaults.dictionary(forKey: storageKey) as? [String: String] ?? [:]
        if let stored = pins[serial] {
            return stored == fingerprint ? .matched : .mismatch
        }
        pins[serial] = fingerprint
        defaults.set(pins, forKey: storageKey)
        return .firstUse
    }

    func delete(for serial: String) {
        lock.lock()
        defer { lock.unlock() }
        let defaults = BambuDefaults.shared
        var pins = defaults.dictionary(forKey: storageKey) as? [String: String] ?? [:]
        pins.removeValue(forKey: serial)
        defaults.set(pins, forKey: storageKey)
    }
}

File: windows/Gantry.Windows/Services/CertificatePinStore.cs
using System.Security.Cryptography;

namespace Gantry.Services;

/// <summary>Pins each printer's TLS certificate after first use, ported from the macOS store.</summary>
public static class CertificatePinStore
{
    public enum ValidationResult
    {
        FirstUse,
        Matched,
        Mismatch
    }

    public static bool Accepted(this ValidationResult result) => result != ValidationResult.Mismatch;

    private static readonly object Gate = new();
    private const string Key = "printer-certificate-pins-v1";

    public static ValidationResult Validate(byte[] certificateData, string serial)
    {
        string fingerprint = Convert.ToHexString(SHA256.HashData(certificateData)).ToLowerInvariant();
        lock (Gate)
        {
            var pins = Defaults.GetStringDictionary(Key);
            if (pins.TryGetValue(serial, out var stored))
                return stored == fingerprint ? ValidationResult.Matched : ValidationResult.Mismatch;
            pins[serial] = fingerprint;
            Defaults.SetStringDictionary(Key, pins);
            return ValidationResult.FirstUse;
        }
    }

    public static void Delete(string serial)
    {
        lock (Gate)
        {
            var pins = Defaults.GetStringDictionary(Key);
            if (pins.Remove(serial)) Defaults.SetStringDictionary(Key, pins);
        }
    }
}
